### PR TITLE
Change the logic to handle the future map loading

### PR DESCRIPTION
**Pull request by ihhub**
- add checks for objects placement on the Adventure Map
- add abort logic for history manager to revert changes back to the previous state if there is an error
- add logic to load objects on the Adventure Map for the future map loading

relates to #6845

### CODE REVIEW

File: src/fheroes2/editor/history_manager.h [hunk 1]
```diff
@@ -58,10 +58,17 @@ namespace fheroes2
 
         ActionCreator & operator=( const ActionCreator & ) = delete;
 
+        void abort()
```
Comment (oleg-derevenetz) [inline src/fheroes2/editor/history_manager.h:61]: Hi @ihhub what do you think about the reverse approach: add `commit()` method which should be called when everything is done properly, and abort in destructor if `commit()` wasn't called?

Comment (ihhub) [reply src/fheroes2/editor/history_manager.h:61]: Yup, it makes more sense. I shall change it.

**Review by Districh-ru — APPROVED** [review]

File: src/fheroes2/editor/history_manager.cpp [hunk 2]
```diff
@@ -110,11 +110,11 @@ namespace fheroes2
...
-    ActionCreator::~ActionCreator()
+    void ActionCreator::commit()
     {
         auto * action = dynamic_cast<MapAction *>( _action.get() );
         if ( action == nullptr ) {
-            // How is it even possible?
+            // How is it even possible? Did you call this method twice?
             assert( 0 );
             return;
         }
```
Comment (oleg-derevenetz) [inline src/fheroes2/editor/history_manager.cpp:120]: Why do we need the `try`/`catch` block below with just an `assert()` call in `catch`?

Comment (ihhub) [reply src/fheroes2/editor/history_manager.cpp:120]: I've removed it.

**Pull request merged**